<a href="https://colab.research.google.com/github/andreysnigirev/Data_Analysis/blob/main/hw03_eda.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Домашнее задание 3 (10 баллов)

## НИУ ВШЭ, 2024-25 учебный год

## Формат сдачи

Сам ноутбук называйте в формате hw-03-USERNAME.ipynb, где USERNAME — ваши фамилия и имя. Файл надо будет сдать в anytask.

## О задании

В этом домашнем задании вы получите опыт подготовки данных и EDA (Exploratory Data Analysis), приблизившись к реальной работе с данными с помощью уже знакомых вам инструментов: numpy, pandas, matplotlib и seaborn, а также познакомитесь с plotly.

EDA - важный этап работы с данными. Именно в процессе EDA вы можете выяснить особенности датасета, найти зависимости в данных, протестировать гипотезы о взаимосвязях признаков и целевой переменной. В реальности вы сами будете определять, что и как проверять в рамках EDA, однако в данной работе гипотезы будут выдвинуты за вас автором задания. Также и типы графиков, которые вы будете использовать уже предопределены за вас. Выполняя задания по визуализациям, обращайте внимание на референсные картинки, которые были оставлены авторами. Они показывают, какой примерно результат от вас ожидается.

Это задание будет включать в себя построение некоторого количества графиков, потому вот основные правила, в соответствии с которыми графики должны быть оформлены:
1. Любой график всегда имеет название.
2. Оси у графика подписаны (исключение - график делит название оси с соседним).
3. Подписи делаются на одном языке (либо английский, либо русский), но не на их смеси.
4. Если вы выводите несколько графиков, то они размещаются в рамках одного объекта matplotlib.pyplot.figure. Вы не выводите графики последовательно, а смотрите, как их разместить сразу все за раз.
5. Отдельные графики (если графиков на картинке больше 1) также имеют свои названия.
6. Если на графике есть несколько различных объектов, то для их различения используется легенда. Помещайте легенду в место, где она в наименьшей мере мешает восприятию самого графика.
7. Если есть легенда, то она должна иметь навание, а также ее содержание должно быть адекватными (пожалейте проверяющих, они не хотят разбирать, а к чему относится зеленая линия, которая на легенде помечена меткой "0").
8. Используйте цветовые гаммы, чтобы объекты на графиках отличались цветом.
9. Текст на визуализации должен быть читаем. Если вы вывели названия категорий в качестве подписей на оси Х, а эти названия наезжают друг на друга, то вам нужно что-то с этим сделать. Нечитаемый или плохо читаемый (очень маленький, невероятно огромный) текст ведет к снижению баллов.

При работе с табличными данными используйте функционал [style](https://pandas.pydata.org/docs/user_guide/style.html#Builtin-Styles), чтобы визуально выделять необходимые элементы.

В этом задании будут оцениваться ваши развернутые комментарии, размышления и выводы (по построенным графикам): не скупитесь на них, но следите, чтобы выводы не были поверхностными и очевидными.

*Тут правило про плагиат, которое вам итак известно - обнуляются все работы, которые подозреваются в некорректных заимствованиях!* Помните, что просто скопировать код из интернета и указать источник - это не ваша самостоятельная работа, а автора кода. Измените хоть что-нибудь в коде.

## О данных

Будем работать с [датасетом физической активности](https://www.kaggle.com/datasets/uciml/human-activity-recognition-with-smartphones).

В этом датасете собраны данные 30 человек, выполняющих различные действия со смартфоном на поясе. Данные записывались с помощью датчиков (акселерометра и гироскопа) в этом смартфоне. Были зафиксированы: "3-осевое линейное ускорение" (_tAcc-XYZ_) и "3-осевая угловая скорость" (_tGyro-XYZ_).

1. Все сигналы отсортированы по времени, пропущены через фильтр шума.
2. Все наблюдения были разделены на окна по 128 сигналов в каждом, причем окна пересекаются на 50% (грубо говоря, с 1 по 128 сигнал в 1 окно, с 64 до 192 сигнала - во 2 окно и так далее). То есть каждое окно покрывает 2.56 секунды данных (два шага человека при ходьбе, учитывая медленную походку пожилых людей).
3. Линейное ускорение было разделено на две составляющие (с помощью низкочастотного фильтра): ускорение самого тела, а также гравитационная составляющая. Будем называть их ускорением тела и гравитационным ускорением.
4. К получившимся окнам применяли различные функции для получения следующих показателей:
    - max
    - min
    - mad (median)
    - mean
    - std
    - skewness (коэффициент асимметричности распределения признака в окне)
    - etc. (подробнее, на сколько это возможно, указано в [самой статье](https://www.esann.org/sites/default/files/proceedings/legacy/es2013-84.pdf))
5. Величина (magnitude) сигнала также определялась через евклидову норму всех значений из вектора для одного окна.
6. К окнам было применено преобразование Фурье, чтобы получить частоты. Из частот были извелченыте же показатели, плюс новые:
    - bandsEnergy (энергия частотного интервала)
7. Были посчитаны углы между векторами соответствующих окон.
8. Для ускорений были посчитаны производные - рывки, к которым применялись те же функции.

Итог: в датасете каждый признак - это какая-то операция над вектором из 128 чисел, которые соответствуют замерам определенного человека занятого какой-то активностью на протяжении 2.56 секунд.

Как читать названия признаков? Примеры:
- `tBodyAccMag-mean()` - это вектор ускорений тела (без гравитационной составляющей этого ускорения), из которого взяли евклидову норму, а затем усреднили, чтобы получить скаляр
- `fBodyAcc-bandsEnergy()-1,8.2` - это вектор ускорения тела, к которому применили преобразование Фурье, преобразовав его в вектор частот, а затем посчитали на этом векторе энергию данного частотного интервала в определенном диапазоне частот (в данном случае 1 - 8.2)
- `angle(X,gravityMean)` - это угол между осью Х и вектором усредненных значений гравитационного ускорения из соответствующего окна

Вероятно, сначала покажется сложным, но мы разберемся с некоторыми признаками в отдельности и вы сможете понять их смысл, уже исходя из фактических соотношений.

## Предобработка данных

In [85]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings("ignore")

In [86]:
train = pd.read_csv('train.csv')
test = pd.read_csv("test.csv")

In [5]:
train.shape

(3041, 563)

В датасете 563 признака, включая целевую переменную. Столько признаков нам не нужно. Мы за вас сделали подвыборку признаков для удобства работы с ними.

In [84]:
cols_subset = [
    'tBodyAccMag-mean()',  # усредненная величина ускорения тела
    'tBodyGyroJerk-mad()-X',  # медианная величина рывка тела по оси Х
    'tGravityAcc-min()-X',  # минимум гравитационной составляющей ускорения по оси Х
    'tBodyAcc-max()-X',  # максимальная величина ускорения тела по оси Х
    'fBodyAcc-bandsEnergy()-1,8.2',  # энергия ускорения тела в интервале частоты
    'angle(X,gravityMean)',  # угол между осью Х и усредненной гравитационной составляющей ускорения
    'angle(Y,gravityMean)',  # угол между осью Y и усредненной гравитационной составляющей ускорения
    'angle(Z,gravityMean)',  # угол между осью Z и усредненной гравитационной составляющей ускорения,
    'fBodyAcc-skewness()-X',  # асимметричность частоты ускорения тела по оси Х
    'subject',  # номер испытуемого
    'Activity',  # название вида деятельности (целевая переменная)
]

In [83]:
train = train[cols_subset]
test = test[cols_subset]

### Задание 1 (0.5 балла)

Посмотрите на пропуски и дубли в данных train и test. Есть ли они? Если есть, напишите, с чем это может быть связано? Если нет, то просто напишите, что пропусков и/или дублей нет.

In [17]:
train.isna().sum() #при помощи комбинаций функций isna() и sum() найдем количество пропущенных значений в каждом столбце датафрейма train
'''видим, что в каждом столбце, кроме tGravityAcc-min()-X и tBodyAcc-max()-X
есть одно пропущенное значение. (если посмотреть на фрейм, то все они пропущены в последней строке датафрейма)
так как пропущено значение столбца с человеком, на котором производились измерения и данных об активности
можно сделать вывод, что в таблицу эта строка попала по ошибке и подобного измерения на самом деле не было'''

tBodyAccMag-mean()              1
tBodyGyroJerk-mad()-X           1
tGravityAcc-min()-X             0
tBodyAcc-max()-X                0
fBodyAcc-bandsEnergy()-1,8.2    1
angle(X,gravityMean)            1
angle(Y,gravityMean)            1
angle(Z,gravityMean)            1
fBodyAcc-skewness()-X           1
subject                         1
Activity                        1
dtype: int64

In [44]:
test.isna().sum() #при помощи комбинаций функций isna() и sum() найдем количество пропущенных значений в каждом столбце датафрейма train
#видно, что в датафрейме test нет пропущенных значений

tBodyAccMag-mean()              0
tBodyGyroJerk-mad()-X           0
tGravityAcc-min()-X             0
tBodyAcc-max()-X                0
fBodyAcc-bandsEnergy()-1,8.2    0
angle(X,gravityMean)            0
angle(Y,gravityMean)            0
angle(Z,gravityMean)            0
fBodyAcc-skewness()-X           0
subject                         0
Activity                        0
dtype: int64

In [37]:
train.nunique() #определяем количество уникальных значений в каждом столбце train
# видно, что в столбцах tBodyAccMag-mean() и tBodyGyroJerk-mad()-X, angle(X,gravityMean),
# angle(Y,gravityMean), angle(Z,gravityMean), fBodyAcc-skewness()-X повторяющихся значений нет, т.к число
# уникальных элементов в них равно числу строк. В столбце subject столь малое число уникальных элементов обусловлено тем,
# что в данном датафрейме представлены данные об активности только 10 разных субъектов,
# малое число уникальных элементов в столбце activity обусловлено тем, что мы у нас только 6 рассматриваемых активностей.
# наличие повторов в остальных строках может быть вызвано тем, что разные люди могут ходить схожим образом в нескольких случаях,
# таким образом показатели физической активности будут одинаковы в нескольких ситуациях

tBodyAccMag-mean()              3040
tBodyGyroJerk-mad()-X           3040
tGravityAcc-min()-X             2313
tBodyAcc-max()-X                2170
fBodyAcc-bandsEnergy()-1,8.2    3033
angle(X,gravityMean)            3040
angle(Y,gravityMean)            3040
angle(Z,gravityMean)            3040
fBodyAcc-skewness()-X           3040
subject                           10
Activity                           6
dtype: int64

In [43]:
test.nunique() #определяем количество уникальных значений в каждом столбце train
# видно, что в столбцах tBodyAccMag-mean(), angle(X,gravityMean),
# angle(Y,gravityMean), angle(Z,gravityMean), fBodyAcc-skewness()-X повторяющихся значений нет, т.к число
# уникальных элементов в них равно числу строк. В столбце subject столь малое число уникальных элементов обусловлено тем,
# что в данном датафрейме представлены данные об активности только 9 разных субъектов,
# малое число уникальных элементов в столбце activity обусловлено тем, что мы у нас только 6 рассматриваемых активностей.
# наличие повторов в остальных строках может быть вызвано тем, что разные люди могут ходить схожим образом в нескольких случаях,
# таким образом показатели физической активности будут одинаковы в нескольких ситуациях

tBodyAccMag-mean()              2947
tBodyGyroJerk-mad()-X           2946
tGravityAcc-min()-X             2281
tBodyAcc-max()-X                2082
fBodyAcc-bandsEnergy()-1,8.2    2940
angle(X,gravityMean)            2947
angle(Y,gravityMean)            2947
angle(Z,gravityMean)            2947
fBodyAcc-skewness()-X           2947
subject                            9
Activity                           6
dtype: int64

### Задание 2 (0.5 балла)

Выведите основные статистики данных: среднее, медиана, перцентили (0.25, 0.5, 0.75), минимум, максимум, стандартное отклонение. Посмотрите на таблицу и напишите, видите ли вы в ней что-нибудь интересное.

In [48]:
train.describe()

tBodyAccMag-mean()  tBodyGyroJerk-mad()-X  tGravityAcc-min()-X  \
count         3040.000000            3040.000000          3041.000000   
mean            -0.502183              -0.697068             0.709892   
std              0.478657               0.323844             0.463909   
min             -0.999504              -0.999783            -0.710240   
25%             -0.978414              -0.989145             0.809531   
50%             -0.647276              -0.859873             0.927439   
75%             -0.070563              -0.444124             0.965613   
max              0.510146               0.827351             1.000000   

       tBodyAcc-max()-X  fBodyAcc-bandsEnergy()-1,8.2  angle(X,gravityMean)  \
count       3041.000000                   3040.000000           3040.000000   
mean          -0.428288                     -0.821925             -0.514042   
std            0.546669                      0.281652              0.473491   
min           -0.971348                     -0.999996             -1.000000   
25%           -0.933942                     -0.999246             -0.804155   
50%           -0.772358                     -0.932530             -0.699986   
75%            0.001681                     -0.756887             -0.517712   
max            0.784666                      1.000000              0.799174   

       angle(Y,gravityMean)  angle(Z,gravityMean)  fBodyAcc-skewness()-X  \
count           3040.000000           3040.000000            3040.000000   
mean               0.056532             -0.016953              -0.157632   
std                0.333771              0.248386               0.380276   
min               -1.000000             -0.980143              -0.835831   
25%                0.020284             -0.089690              -0.487522   
50%                0.202395              0.009985              -0.199065   
75%                0.262375              0.108567               0.111714   
max                0.385117              0.488889               0.963346   

           subject  
count  3040.000000  
mean      8.176316  
std       4.837589  
min       1.000000  
25%       5.000000  
50%       7.000000  
75%      14.000000  
max      16.000000

In [49]:
test.describe()

tBodyAccMag-mean()  tBodyGyroJerk-mad()-X  tGravityAcc-min()-X  \
count         2947.000000            2947.000000          2947.000000   
mean            -0.559043              -0.735661             0.696878   
std              0.439539               0.282074             0.502873   
min             -0.998936              -1.000000            -0.756774   
25%             -0.978433              -0.989078             0.836787   
50%             -0.861830              -0.914698             0.934930   
75%             -0.149596              -0.486423             0.970871   
max              0.479547               0.117606             0.996148   

       tBodyAcc-max()-X  fBodyAcc-bandsEnergy()-1,8.2  angle(X,gravityMean)  \
count       2947.000000                   2947.000000           2947.000000   
mean          -0.462063                     -0.880361             -0.513923   
std            0.523916                      0.160496              0.509205   
min           -0.952357                     -0.999989             -0.984195   
25%           -0.934447                     -0.999330             -0.829722   
50%           -0.852659                     -0.979336             -0.729648   
75%           -0.009965                     -0.799113             -0.545939   
max            0.786436                      0.048730              0.833180   

       angle(Y,gravityMean)  angle(Z,gravityMean)  fBodyAcc-skewness()-X  \
count           2947.000000           2947.000000            2947.000000   
mean               0.074886             -0.048720              -0.154042   
std                0.324300              0.241467               0.392203   
min               -0.913704             -0.949228              -1.000000   
25%                0.022140             -0.098485              -0.497070   
50%                0.181563             -0.010671              -0.202050   
75%                0.260252              0.092373               0.140254   
max                1.000000              0.973113               0.960367   

           subject  
count  2947.000000  
mean     12.986427  
std       6.950984  
min       2.000000  
25%       9.000000  
50%      12.000000  
75%      18.000000  
max      24.000000

### Задание 3 (0.5 балла)

Данные сбалансированы? У вас есть 6 категорий активности (целевая переменная). Если одна из категорий представлена большим количеством наблюдений, а остальные - нет, то это обстоятельство, о котором хотелось бы знать заранее. Гиперболизированный пример: представьте, что у вас 1000 наблюдений, разделенных на два класса, и только 5 принадлежат ко второму классу. Как учить модель на такой выборке? Есть сильное подозрение, что многие модели будут склонны вырождаться в константу. Отсюда и необходимость проверять баланс классов.

Выведите количество наблюдений для всех категорий активности в train и test. Посчитайте разницу в процентах между самой крупной категорией и самой маленькой (самая большая категория - это 100%). Если разница (округленная до ближайшего целого числа) не превышает 30%, то давайте для наших целей считать, что данные сбалансированы. Напишите, сбалансированы ли данные в train и test.

In [72]:
#так как функция count() будет возвращать одно и то же значение для каждого столбца при конкретном значении Activity,
#то нам не важно какой из столцов брать. (Возьмем для подсчетов столбец subject)
train.groupby("Activity")["subject"].count()
#количество наблюдений для всех категорий активности в датафрейме train

Activity
LAYING                542
SITTING               489
STANDING              530
WALKING               569
WALKING_DOWNSTAIRS    435
WALKING_UPSTAIRS      475
Name: subject, dtype: int64

In [69]:
test.groupby("Activity")["subject"].count()
#количество наблюдений для всех категорий активности в датафрейме train

Activity
LAYING                537
SITTING               491
STANDING              532
WALKING               496
WALKING_DOWNSTAIRS    420
WALKING_UPSTAIRS      471
Name: subject, dtype: int64

In [73]:
round(100 - test.groupby("Activity")["subject"].count().min() / (test.groupby("Activity")["subject"].count().max() / 100))
#разница меньше 30% значит данные в датафрейме test сбалансированы

22

In [74]:
round(100-train.groupby("Activity")["subject"].count().min() / (train.groupby("Activity")["subject"].count().max() / 100))
#разница меньше 30% значит данные в датафрейме train сбалансированы

24

### Задание 4 (0.5 балла)

Посчитайте корреляции всех признаков (включая таргет) со всеми другими. Какими получились значения? Есть идеи, почему так? Напишите ваши мысли.

In [92]:
#при подсчете корреляции вознакет проблема, т.к в столбце Activity данные представлены в строковом виде,
#поэтому, для подсчета корреляции необходимо перевести данные из столбца Activity в числовой тип данных
#для этого рационально использовать LabelEncoder создадим глубокую копию, с которой будем работать,
#чтобы не менять данные в исходной таблице
from sklearn.preprocessing import LabelEncoder
label_encoder = LabelEncoder()
test_copy = test.copy(deep = True)
test_copy["Activity"] = label_encoder.fit_transform(test_copy["Activity"])
test_copy.corr()

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  \
tBodyAcc-mean()-X              1.000000           0.041274          -0.129645   
tBodyAcc-mean()-Y              0.041274           1.000000           0.225980   
tBodyAcc-mean()-Z             -0.129645           0.225980           1.000000   
tBodyAcc-std()-X               0.016984          -0.054264          -0.038578   
tBodyAcc-std()-Y              -0.001799          -0.059066          -0.048340   
...                                 ...                ...                ...   
angle(X,gravityMean)          -0.058421          -0.017138          -0.013933   
angle(Y,gravityMean)           0.034220          -0.030253          -0.007318   
angle(Z,gravityMean)           0.038936          -0.027410          -0.051057   
subject                        0.005077           0.003163           0.021476   
Activity                       0.019516          -0.071542          -0.085756   

                      tBodyAcc-std()-X  tBodyAcc-std()-Y  tBodyAcc-std()-Z  \
tBodyAcc-mean()-X             0.016984         -0.001799         -0.008065   
tBodyAcc-mean()-Y            -0.054264         -0.059066         -0.077051   
tBodyAcc-mean()-Z            -0.038578         -0.048340         -0.042342   
tBodyAcc-std()-X              1.000000          0.910636          0.896031   
tBodyAcc-std()-Y              0.910636          1.000000          0.874501   
...                                ...               ...               ...   
angle(X,gravityMean)         -0.382696         -0.383742         -0.383490   
angle(Y,gravityMean)          0.401433          0.467572          0.427469   
angle(Z,gravityMean)          0.388747          0.405681          0.488541   
subject                      -0.068487         -0.036466         -0.028508   
Activity                      0.821475          0.809545          0.829412   

                      tBodyAcc-mad()-X  tBodyAcc-mad()-Y  tBodyAcc-mad()-Z  \
tBodyAcc-mean()-X             0.022942         -0.002124         -0.010085   
tBodyAcc-mean()-Y            -0.052501         -0.062012         -0.081834   
tBodyAcc-mean()-Z            -0.037851         -0.049705         -0.037628   
tBodyAcc-std()-X              0.998828          0.904884          0.890581   
tBodyAcc-std()-Y              0.909197          0.997942          0.873176   
...                                ...               ...               ...   
angle(X,gravityMean)         -0.381896         -0.380548         -0.387125   
angle(Y,gravityMean)          0.397630          0.470157          0.428295   
angle(Z,gravityMean)          0.386152          0.406818          0.484416   
subject                      -0.064529         -0.036954         -0.039018   
Activity                      0.814042          0.810608          0.825336   

                      tBodyAcc-max()-X  ...  fBodyBodyGyroJerkMag-kurtosis()  \
tBodyAcc-mean()-X             0.055897  ...                         0.036449   
tBodyAcc-mean()-Y            -0.047736  ...                         0.003612   
tBodyAcc-mean()-Z            -0.051682  ...                         0.035943   
tBodyAcc-std()-X              0.983415  ...                         0.189835   
tBodyAcc-std()-Y              0.894876  ...                         0.227084   
...                                ...  ...                              ...   
angle(X,gravityMean)         -0.384454  ...                        -0.084848   
angle(Y,gravityMean)          0.410477  ...                         0.075210   
angle(Z,gravityMean)          0.398008  ...                         0.056870   
subject                      -0.071786  ...                        -0.013617   
Activity                      0.829569  ...                         0.084340   

                      angle(tBodyAccMean,gravity)  \
tBodyAcc-mean()-X                       -0.582641   
tBodyAcc-mean()-Y                        0.112800   
tBodyAcc-mean()-Z                        0.064029   
tB

Поздравляем! Вы предобработали данные для максимально комфортного их последующего анализа. Теперь вы готовы приступить к Exploratory Data Analysis!

In [93]:
train_copy = train.copy(deep = True)
train_copy["Activity"] = label_encoder.fit_transform(train_copy["Activity"])
train_copy.corr()

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  \
tBodyAcc-mean()-X              1.000000           0.148061          -0.256952   
tBodyAcc-mean()-Y              0.148061           1.000000          -0.078769   
tBodyAcc-mean()-Z             -0.256952          -0.078769           1.000000   
tBodyAcc-std()-X               0.000619          -0.045160          -0.020217   
tBodyAcc-std()-Y              -0.021903          -0.044920          -0.016641   
...                                 ...                ...                ...   
angle(X,gravityMean)          -0.035257          -0.005309           0.008587   
angle(Y,gravityMean)           0.034371           0.001053          -0.015288   
angle(Z,gravityMean)           0.028242          -0.013903          -0.022643   
subject                        0.024181          -0.003144          -0.000637   
Activity                       0.005689          -0.062310          -0.054524   

                      tBodyAcc-std()-X  tBodyAcc-std()-Y  tBodyAcc-std()-Z  \
tBodyAcc-mean()-X             0.000619         -0.021903         -0.044617   
tBodyAcc-mean()-Y            -0.045160         -0.044920         -0.049746   
tBodyAcc-mean()-Z            -0.020217         -0.016641         -0.008410   
tBodyAcc-std()-X              1.000000          0.927461          0.851668   
tBodyAcc-std()-Y              0.927461          1.000000          0.895510   
...                                ...               ...               ...   
angle(X,gravityMean)         -0.371653         -0.380531         -0.345011   
angle(Y,gravityMean)          0.471065          0.523600          0.476006   
angle(Z,gravityMean)          0.394825          0.433169          0.482828   
subject                      -0.064345         -0.115524         -0.050123   
Activity                      0.815583          0.825453          0.802000   

                      tBodyAcc-mad()-X  tBodyAcc-mad()-Y  tBodyAcc-mad()-Z  \
tBodyAcc-mean()-X             0.006290         -0.022754         -0.047558   
tBodyAcc-mean()-Y            -0.044180         -0.045049         -0.050402   
tBodyAcc-mean()-Z            -0.018747         -0.015203         -0.001988   
tBodyAcc-std()-X              0.998632          0.920888          0.846392   
tBodyAcc-std()-Y              0.922803          0.997347          0.894509   
...                                ...               ...               ...   
angle(X,gravityMean)         -0.368191         -0.377025         -0.347389   
angle(Y,gravityMean)          0.466424          0.525081          0.477607   
angle(Z,gravityMean)          0.390922          0.431459          0.479751   
subject                      -0.063440         -0.114753         -0.055457   
Activity                      0.806112          0.825999          0.805568   

                      tBodyAcc-max()-X  ...  fBodyBodyGyroJerkMag-kurtosis()  \
tBodyAcc-mean()-X             0.044062  ...                        -0.015613   
tBodyAcc-mean()-Y            -0.038108  ...                        -0.004459   
tBodyAcc-mean()-Z            -0.037197  ...                         0.019127   
tBodyAcc-std()-X              0.980844  ...                         0.115565   
tBodyAcc-std()-Y              0.917366  ...                         0.176946   
...                                ...  ...                              ...   
angle(X,gravityMean)         -0.384192  ...                        -0.079751   
angle(Y,gravityMean)          0.480229  ...                         0.078079   
angle(Z,gravityMean)          0.405023  ...                         0.052548   
subject                      -0.055633  ...                        -0.043902   
Activity                      0.819679  ...                         0.056951   

                      angle(tBodyAccMean,gravity)  \
tBodyAcc-mean()-X                       -0.544320   
tBodyAcc-mean()-Y                        0.070559   
tBodyAcc-mean()-Z                        0.052841   
tB

## Exploratory Data Analysis

### Задание 5 (0.5 балла)

Вы проверили основную информацию о данных и составили свое первое впечатление о них. Теперь визуализации.

Начнем рисовать графики с небольшой разминки. Выберите две любые математические функции (отличные от линейной) и нарисуйте их на одном графике. Подпишите каждый из графиков, чтобы было понятно, какие функции вы выбрали. Ниже будет пример.

In [109]:
# (◕‿◕)♡ your code here

### Задание 6 (0.5 балла)

Ранее мы уже проверили сбалансированность данных по разным категориям активностей в train и test сетах. Теперь визуализируйте количество сэмплов в каждой из категорий отдельно для train и test сетов, как в примере ниже. Проследите, чтобы порядок категорий слева и справа был одинаков. Иначе не очень удобно их сравнивать (когда они в разных местах двух графиков). Также заметьте, что в примере ось Х в одинаковом масштабе, что тоже удобно для сравнения.

Вам может пригодиться:
- plt.barh или ax.barh

In [110]:
# ̿' ̿'\̵͇̿̿\з=( ͡ °_̯͡° )=ε/̵͇̿̿/'̿'̿ ̿ your code here

Правда на таких графиках проще проверять баланс классов таргета?)

### Задание 7 (0.5 балла)

Ранее вы выводили таблицу корреляций. Однако просто таблица корреляций не очень удобна для анализа. Что, если признаков было бы больше? Визуализируйте heatmap корреляций. Подпишите значения корреляций, округленные до 2 знаков после запятой, на графике и оставьте только те значения, которые по модулю не меньше 0.5. Будем считать такие корреляции существенными. Если будете брать код из семинарского ноутбука, то поменяйте в нем что-нибудь, например, цветовую гамму, иначе проверяющий будет карать :)

Какой признак стоит исключить из рассмотрения корреляций? Напишите его название, а также причину, почему нет смысла его рассматривать.

Вам может пригодиться:
- sns.heatmap
- методы get_text и set_text у атрибута texts графика

In [111]:
# ღゝ◡╹)ノ♡ your code here

### Задание 8 (0.5 балла)

Для начала будем стрелять из пушки по воробьям. Выведите pairplot по всем данным train сета. Напишите, заметили ли вы в нем что-нибудь интересное, чего не было в таблицах и графиках ранее, а также сообщите, было ли вам удобно анализировать 100 графиков.

In [112]:
# ♪ヽ( ⌒o⌒)人(⌒-⌒ )v ♪ your code here

### Задание 9 (1 балл)

Окей. Смотреть на все сразу не очень удобно. Кажется, нужно смотреть на признаки в отдельности и искать в них инсайты.

Попробуем проверить гипотезу. Отобразите плотность распределения (kde) `tBodyAccMag-mean()` для каждой из активностей, как в примере ниже. Видите закономерность? Можно ли поделить все активности на две группы? Если да, то проведите вертикальную линию, которая визуально неплохо разделяет эти группы.

Давайте первую группу называть статическими активностями, а вторую - динамическими.

Посчитайте, какую долю правильных ответов дает ваша вертикальная линия для разделения именно этих двух групп. С чем может быть связан полученный вами результат? Предположите логическое обоснование (почему именно данный признак так хорошо/плохо разделяет группы активностей).

Вам может пригодиться:
- sns.kdeplot

In [113]:
# ʅ(◔౪◔ ) ʃ your code here

### Задание 10 (0.5 балла)

Поздравляем, вы проверили первую гипотезу в процессе EDA. Однако нам бы хотелось уметь не группы разделять, а каждый отдельный вид активности. Попробуйте нарисовать "ящики с усами" для `angle(X,gravityMean)` с разделением по разным видам активности на данных train. Можно ли сделать какой-то вывод по полученной визуализации? Проанализируйте и проинтерпретируйте полученные картинки (если заметили какую-то особенность, может какая-то категория значительно выделяется, то предположите, почему так могло произойти).

Вам может пригодиться:
- sns.boxplot

In [114]:
# ^⨀ᴥ⨀^ your code here

### Задание 11 (0.5 балла)

Посмотрим теперь на `tBodyAcc-max()-X`. Визуализируйте скрипичные диаграммы для этого признака с разделением по активностям на всем train. Кажется логичным, что максимальное ускорение в моменте для разных активностей будет различаться. Однако есть ли активность, которая будет выделяться? Если да, то предложите объяснение, почему эта активность выделяется.

Вам может пригодиться:
- sns.violinplot

In [115]:
# ( ͡☉ ͜ʖ ͡☉) your code here

### Задание 12 (1 балл)

Вы уже можете с довольно хорошим качеством выделить две категории активности из шести. Продолжим.

Ранее уже упоминалось разделение на две группы активностей: статические и динамические. Пришла пора разделить датасет train на две части. Сделайте это! Теперь на одной картинке покажите распределения `fBodyAcc-bandsEnergy()-1,8.2` для динамичных активностей (`boxenplot` - немного измененый "ящик с усами") и `angle(Y,gravityMean)` для статических - тут хватит и простой гистограммы (прямо, как в примере ниже). Можно ли выделить еще какие-то активности? Если да, напишите, какие активности на 1 и 2 графиках выделяются.

Вам может пригодиться:
- plt.subplots или matplotlib.gridspec (заметили, что один график больше другого? попробуйте сами так сделать - вам понравится...)
- sns.boxenplot
- sns.histplot

In [116]:
# ‿( ́ ̵ _-`)‿ your code here

### Задание 13 (0.5 балла)

Не всегда жизнь будет такой легкой, а данные будут полностью или почти линейно разделимы в одному признаку (иначе чем бы занимались дата саентисты...). Часто данные вообще ни по одному признаку нельзя адекватно разделить (да и по комбинациям признаков тоже). Давайте попробуем взять не один признак, а целых два!

Рассмотрим только статические активности. Будет снова два графика сразу. На первом графике будет боксплот распределения `fBodyAcc-skewness()-X` по активностям. А на втором попробуем взять не все наблюдения данной группы активностей, а только их часть. Разделимы ли категории по `fBodyAcc-skewness()-X`? А если взять только те строки, где `fBodyAcc-skewness()-X < -0.376`, и теперь вывести боксплот распределения `angle(Y,gravityMean)` по разным активностям? Попробуйте и напишите, выделилась ли какая-то категория?

Вам может пригодиться:
- plt.subplots или matplotlib.gridspec
- sns.boxplot

In [117]:
# (ಠ_ಠ) your code here

Поздравляем, теперь вы на шажок ближе к тому, чтобы стать решающим деревом!

### Задание 14 (1 балл)

Вот и осталась лишь одна активность, которую мы никак не выделили. Попробуем и ее какими-то порогами отделить от всех остальных. Да, можно было бы сказать, что, раз уж мы выделили 5 из 6 категорий, то все остальное - это и есть последняя категория, но мы не идем легкими путями. Выделить оставшуюсь категорию с хорошим качеством не так уж и просто. Из доступных признаков, сначала посмотрите на распределение `tBodyAcc-max()-X` на динамических активностях (скрипичная диаграмма отлично подойдет).

А теперь спускаем вам порог `tBodyAcc-max()-X < 0.157`. Для таких строк постройте диаграмму рассеяния для признака `tGravityAcc-min()-X`. Да, получается неидеально, но это честная работа. Можно выбрать порог так, чтобы чисто из таблицы с динамическими активностями из строк, где `tBodyAcc-max()-X < 0.157`, выбрать по некоторому порогу для `tGravityAcc-min()-X` строки так, чтобы доля правильных ответов на оставшейся категории была не меньше 83%!

Вам может пригодиться:
- plt.subplots или matplotlib.gridspec
- sns.violinplot
- sns.swarmplot (иногда лучше подходит, чем просто sns.scatterplot)

Обратите внимание, что подписи к осям можно убирать методами plt. Например, если у вас общее название для оси Y для обоих графиков, то можно не подписывать ось Y у второго графика.

In [118]:
# (✖╭╮✖) your code here

### Задание 15 (0.5 балла)

Теперь попробуем более интерактивные графики. Установите библиотеку `plotly` (обычно `pip install plotly` или `conda install plotly`). Синтаксис очень похож на `seaborn`, но вот [документация для plotly.express](https://plotly.com/python-api-reference/plotly.express.html).

Сначала подготовим данные для визуализации. Будем считать t-SNE просто "черным ящиком", который принимает наши данные и проецирует их в двумерное или трехмерное пространство. Возможно, придется немного подождать.

In [ ]:
%%time
from sklearn.manifold import TSNE
tsne = TSNE(random_state=666, n_components=2)
new = tsne.fit_transform(train.drop(['subject', 'Activity'], axis=1))
df_2_comp = pd.concat([
    pd.DataFrame(new, columns=['x', 'y']),
    train['Activity']
], axis=1)

tsne = TSNE(random_state=666, n_components=3)
new = tsne.fit_transform(train.drop(['subject', 'Activity'], axis=1))
df_3_comp = pd.concat([
    pd.DataFrame(new, columns=['x', 'y', 'z']),
    train['Activity']
], axis=1)

Теперь нарисуйте scatter-график через `plotly.express` для `df_2_comp` с выделением цветами активностей. Красиво получилось? Видны скопления точек одного цвета? Можно разные активности отделить линиями на данном графике?

In [ ]:
# ʅ(◔౪◔ ) ʃ your code here

Мы не отстаем от трендов и тоже можем в 3D. Сейчас и вы научитесь (если еще не умеете). Нарисуйте scatter_3d-график через `plotly.express` для `df_3_comp` с выделением цветами активностей. Повращайте его. Нормально получилось? Есть ли какая-то структура у точек?

In [ ]:
# (✧≖‿ゝ≖) your code here

### Задание 16 (1 балл)

Мы сделали вам fit. Ваша очередь сделать predict. Допишите функцию, которая принимает на вход датафрейм с признаками, а выдает вектор из предсказаний - для каждой строки датафрейма вы определяете, к какому классу принадлежит объект. Рекомендуем сразу записывать из строками, а не кодировать в числа. Дальше фукнция `accuracy_score` выдаст вам долю правильных ответов.

Подсказка: вы уже отделили категории друг от друга по каким-то признакам. Воспользуйтесь этим знанием! Используя только проведенный вами анализ можно спокойно выбить не менее 75% на train и test! Если вы получите меньше 75%, то балл будет снижен. Авторское решение (на основе всех графиков выше) выбивает 81% и 77% на train и test соответственно. [Шпаргалка с готовыми идеями для такой задачи](https://rutube.ru/video/f884aa6ed5f94120b7304506042fe5bb/)

In [ ]:
def predict(df: pd.DataFrame):
    # ╰( ͡° ͜ʖ ͡° )つ──☆*:・ﾟ your code here
    pass

def accuracy_score(y_true, y_pred) -> float:
    y_true = np.asarray(y_true).reshape(-1, 1)
    y_pred = np.asarray(y_pred).reshape(-1, 1)
    return (y_true == y_pred).mean()

In [ ]:
train_preds = predict(train)
test_preds = predict(test)
print(f"Accuracy on train set => {accuracy_score(train['Activity'], train_preds)}")
print(f"Accuracy on test set => {accuracy_score(test['Activity'], test_preds)}")

__Место для мема и/или фидбэка__

> тут